# Analyse of Dataset

## Import of Libraries

In [185]:
# use lib pandas para trabalhar com analise de dados
import pandas as pd
# use lib numpy para trabalhar com calculos matematicos dos dados
import numpy as np
# use lib plotly.express para criar mapas.
import plotly.express as px

## Import of Data

In [226]:
# load dataset
df = pd.read_csv('data/kc_house_data.csv')

# read top 5 rows dataset
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

### Check Data Types

In [227]:
# Check data types of dataset
df.dtypes

id                 int64
date              object
price            float64
bedrooms           int64
bathrooms        float64
sqft_living        int64
sqft_lot           int64
floors           float64
waterfront         int64
view               int64
condition          int64
grade              int64
sqft_above         int64
sqft_basement      int64
yr_built           int64
yr_renovated       int64
zipcode            int64
lat              float64
long             float64
sqft_living15      int64
sqft_lot15         int64
dtype: object

### Check NA 

In [228]:
# check NA values
df.isna().sum()

id               0
date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

## ETL Data

### Check duplicated. 

In [229]:
# Check duplicated data from ID columns.
df['id'].duplicated().value_counts()

False    21436
True       177
Name: id, dtype: int64

**OBS:** There are 177 duplicated house ID, may be its happen because there are some update sale price how to show below.

In [230]:
# I used group by to list all houses duplicated
pd.concat(cont for i, cont in df.groupby("id") if len(cont) > 1).head(20)

id             date     price  bedrooms  bathrooms  sqft_living  \
2496     1000102  20140916T000000  280000.0         6       3.00         2400   
2497     1000102  20150422T000000  300000.0         6       3.00         2400   
16814    7200179  20141016T000000  150000.0         2       1.00          840   
16815    7200179  20150424T000000  175000.0         2       1.00          840   
11433  109200390  20140820T000000  245000.0         3       1.75         1480   
11434  109200390  20141020T000000  250000.0         3       1.75         1480   
12417  123039336  20140611T000000  148000.0         1       1.00          620   
12418  123039336  20141208T000000  244900.0         1       1.00          620   
7792   251300110  20140731T000000  225000.0         3       2.25         2510   
7793   251300110  20150114T000000  358000.0         3       2.25         2510   
9234   302000375  20140814T000000  169100.0         3       2.00         1050   
9235   302000375  20150506T000000  250000.0         3       2.00         1050   
14855  324000530  20140708T000000  201500.0         3       1.00         1320   
14856  324000530  20150323T000000  459000.0         3       1.00         1320   
7178   526059224  20140923T000000  260000.0         4       1.75         1650   
7179   526059224  20150206T000000  470000.0         4       1.75         1650   
17381  641900050  20140819T000000  335000.0         4       2.25         2160   
17382  641900050  20150206T000000  499950.0         4       2.25         2160   
19552  643300040  20141104T000000  481000.0         4       1.75         1920   
19553  643300040  20150313T000000  719521.0         4       1.75         1920   

       sqft_lot  floors  waterfront  view  ...  grade  sqft_above  \
2496       9373     2.0           0     0  ...      7        2400   
2497       9373     2.0           0     0  ...      7        2400   
16814     12750     1.0           0     0  ...      6         840   
16815     12750     1.0           0     0  ...      6         840   
11433      3900     1.0           0     0  ...      7        1480   
11434      3900     1.0           0     0  ...      7        1480   
12417      8261     1.0           0     0  ...      5         620   
12418      8261     1.0           0     0  ...      5         620   
7792      12013     2.0           0     0  ...      8        2510   
7793      12013     2.0           0     0  ...      8        2510   
9234      18304     1.0           0     0  ...      7        1050   
9235      18304     1.0           0     0  ...      7        1050   
14855      5000     1.5           0     0  ...      7        1320   
14856      5000     1.5           0     0  ...      7        1320   
7178       7276     1.0           0     0  ...      7        1010   
7179       7276     1.0           0     0  ...      7        1010   
17381      8817     1.0           0     0  ...      7        1460   
17382      8817     1.0           0     0  ...      7        1460   
19552      9500     1.0           0     0  ...      7        1470   
19553      9500     1.0           0     0  ...      7        1470   

       sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
2496               0      1991             0    98002  47.3262 -122.214   
2497               0      1991             0    98002  47.3262 -122.214   
16814              0      1925             0    98055  47.4840 -122.211   
16815              0      1925             0    98055  47.4840 -122.211   
11433              0      1980             0    98023  47.2977 -122.367   
11434              0      1980             0    98023  47.2977 -122.367   
12417              0      1939             0    98106  47.5138 -122.364   
12418              0      1939             0    98106  47.5138 -122.364   
7792               0      1988             0    98003  47.3473 -122.314   
7793               0      1988             0    98003  47.3473 -122.314   
9234               0      1953       

**Remove duplicates from column ID (last)**

In [231]:
# I used drop duplicates to remove and keep only last ID
df.drop_duplicates(subset='id', keep='last', inplace=True)

In [232]:
#check duplicated again
df.duplicated().value_counts()

False    21436
dtype: int64

### Convert data type 

In [233]:
# use o pd.to_datetime para converter as colunas para tipo data.

df['date'] = pd.to_datetime( df['date'])



In [234]:
#df.dtypes
df[['id','date']].head()

id       date
0  7129300520 2014-10-13
1  6414100192 2014-12-09
2  5631500400 2015-02-25
3  2487200875 2014-12-09
4  1954400510 2015-02-18

### 1 - Crie uma nova coluna chamada: "house_age" 
 - Se o valor da coluna "date" for menor que 2015-01-01 <= 'old_house' 
 - Se o valor da coluna "date" for maior que 2015-01-01 >= 'new_house'    

In [235]:
# Use o np.select para criar nova coluna a partir de 2 ou mais condicoes e atribuir os valore dessa nova coluna.

conditionlist = [
    (df['date'] >= '2015-01-01'),
    (df['date'] <= '2015-01-01')]
choicelist = [ 'new_house','old_house']
df['house_age'] = np.select(conditionlist, choicelist, default='not specified')

In [236]:
df.groupby(['house_age']).size().rank

<bound method NDFrame.rank of house_age
new_house     6972
old_house    14464
dtype: int64>

In [237]:
df[['date','house_age']].head()

date  house_age
0 2014-10-13  old_house
1 2014-12-09  old_house
2 2015-02-25  new_house
3 2014-12-09  old_house
4 2015-02-18  new_house

### 2 - Crie uma nova coluna chamada: "dormitory_type" 
- Se o valor da coluna "bedrooms" for igual à 1 => 'studio'
- Se o valor da coluna "bedrooms" for igual a 2 => 'apartament' 
- Se o valor da coluna "bedrooms" for maior que 2 => 'house' 

In [238]:
list1 = [
    (df['bedrooms'] <= 1),
    (df['bedrooms'] == 2),
    (df['bedrooms'] > 2)]
list2 = ['studio','apartment','house']
df['dormitory_type'] = np.select( list1, list2, default='not defined')

In [239]:
# rank de atribuicao do dormitory_type
df.groupby(['dormitory_type']).size().sort_values()

dormitory_type
studio         207
apartment     2736
house        18493
dtype: int64

In [240]:
# testa se todos as linhas tiveram atrubuicao
df[['bedrooms','dormitory_type']].loc [df['dormitory_type'] == 'not defined' ].head()


Empty DataFrame
Columns: [bedrooms, dormitory_type]
Index: []

### 3. Crie uma nova coluna chamada: "condition type" 
- Se o valor da coluna "condition" for menor ou igual à 2 <= 'bad' 
- Se o valor da coluna "condition" for igual à 3 ou 4 = 'regular' 
- Se o valor da coluna "condition" for igual à 5 => 'good' 


In [241]:
list1 = [
    (df['condition'] <= 2),
    (df['condition'] > 2) & (df['condition'] < 5),
    (df['condition'] >= 5)]
list2 = ['bad','regular','good']
df['condition_type'] = np.select( list1, list2, default='not defined')

In [242]:
df.groupby(['condition_type']).size().sort_values()

condition_type
bad          193
good        1687
regular    19556
dtype: int64

In [243]:
df[['condition','condition_type']].loc [df['condition_type'] == 'not defined'].head()

Empty DataFrame
Columns: [condition, condition_type]
Index: []

### 4 - Modifique o TIPO a Coluna "condition" para STRING 

In [244]:
df.astype({'condition': 'str'}).dtypes

id                         int64
date              datetime64[ns]
price                    float64
bedrooms                   int64
bathrooms                float64
sqft_living                int64
sqft_lot                   int64
floors                   float64
waterfront                 int64
view                       int64
condition                 object
grade                      int64
sqft_above                 int64
sqft_basement              int64
yr_built                   int64
yr_renovated               int64
zipcode                    int64
lat                      float64
long                     float64
sqft_living15              int64
sqft_lot15                 int64
house_age                 object
dormitory_type            object
condition_type            object
dtype: object

### 5 - Delete as colunas: "sqft.living15" e "sqft_lot15"  

In [245]:
del df['sqft_living15']

In [246]:
del df['sqft_lot15']

### 6 - Modifique o TIPO a Coluna "yr_build" para DATE 

In [247]:
df['yr_built'] = pd.to_datetime(df['yr_built'] , format= "%Y" )

### 7 -  Modifique o TIPO a Coluna ')"yr_renovated" para DATE 

**nesse caso primeiro e necessario subitituir os valores = 0 por um valor do tipo ano e depois fazer a conversao**
**assim substitui os valor zero como o mesmo valor do anos de contrucao**

In [182]:
# use o np.where para fazer update na coluna 'yr_renovated' onde for 0 preenche com o valor da coluna 'yr_built'.
df['yr_renovated'] = np.where(df['yr_renovated'] == 0, df['yr_built'], df['yr_renovated'] )

TypeError: invalid type promotion

In [213]:
# use o .to_datetime para converter em date
df['yr_renovated'] = pd.to_datetime(df['yr_renovated']  , format= "%Y")

In [214]:
df[['yr_renovated','yr_built']].head()

yr_renovated  yr_built
0 1970-01-01 00:00:00.000000000      1955
1 1970-01-01 00:00:00.000001991      1951
2 1970-01-01 00:00:00.000000000      1933
3 1970-01-01 00:00:00.000000000      1965
4 1970-01-01 00:00:00.000000000      1987

## Answering business questions.

### 8 - Qual a data mais antiga de construção de um imóvel?  

In [175]:
df['yr_built'].min()

1900

**Resposta:** O imovel mais antigo e do ano de 1900.

### 9 -  Qual a data mais antiga de renovação de um imóvel?

In [211]:
df[df['yr_renovated']  '1900-01-01 00:00:00'].min().head()

TypeError: Invalid comparison between dtype=int64 and str

### 10 - Quantos imóveis tem 2 andares? 

In [222]:
# use o ex para listar o numero de floors e suas quantidades no dataset' 
#df.groupby(['floors']).size().sort_values()

# use o ex para listar o quantidades de casas com floor == 2 no dataset'
df[['floors']].loc [df['floors'] == 2].count()

floors    8241
dtype: int64

**Resposta** Existem 8241 casa com 2 floors.

### 11 - Quantos imóveis estão com a condição igual a "regular" ?

In [250]:
df[['condition_type']].loc [df['condition_type'] == 'regular'].count()

condition_type    19556
dtype: int64

**Resposta** Existem 19556 imoveisna condicao regular.

### 12 - Quantos imóveis estão com a condição igual a "bad"e possuem "vista para agua?

In [267]:
# o ex abaixo para filtrar com 2 ou mais condcoes e exibir a quantidade
df[(df['condition_type'] == 'bad') & (df['waterfront'] == 1)].shape[0]



2

**Resposta** Existem 2 imoveis de frente para o mar na condicao bad.

### 13 - Quantos imóveis estão com a condição igual a "good" e são "new house" ?  

In [268]:
df[(df['condition_type'] == 'good') & (df['house_age'] == 'new_house')].shape[0]

421

**Resposta** Existem 421 imoveis com condicao igual good e sao new house.

###  14 - Qual o valor do imóvel mais caro do tipo "studio" ? 

In [272]:
df[['price']].loc [df['dormitory_type'] == 'studio'].max()

price    1295650.0
dtype: float64

**Resposta** O studio mais cara custa 1295650.0

## Answering business questions.

### 1 - Quantas casas estao diponiveis para compra.

In [11]:
# Usei nunique para contar quantidade de chave unica do ID
df['id'].nunique()

21436

**Resposta** Exixtem 21.436 casa disponivel para compra. 

### 2 - Quantos atributos as casas possuem? 

In [12]:
# use o Shape para contar a quantidade de linhas ou colunas do dataset
# df.shape[1] conta as colunas
# df.shape[0] conta as linhas
df.shape[1]

21

**Resposta** As casa possuem 21 atributos.

###  3 - Quais são os atributos das casas?

In [13]:
#use o columns para exibir os nomes das colunas do dataset
df.columns

Index(['id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living',
       'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade',
       'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode',
       'lat', 'long', 'sqft_living15', 'sqft_lot15'],
      dtype='object')

**Resposta** Os atributos sao id, date, price, bedrooms, bathrooms, sqft_living, sqft_lot, floors, waterfront, view,
condition, grade, sqft_above, sqft_basement, yr_built, yr_renovated, zipcode, lat, long, sqft_living15, sqft_lot15

###  4 - Qual a casa mais cara (casa com o maior valor de venda)?

In [14]:
# use o sort_values(by=) para buscar valores ordenados por uma coluna
# pode ser usando tanbem o .rank()
df.sort_values(by='price',ascending=False).head(1)

id             date      price  bedrooms  bathrooms  \
7252  6762700020  20141013T000000  7700000.0         6        8.0   

      sqft_living  sqft_lot  floors  waterfront  view  ...  grade  sqft_above  \
7252        12050     27600     2.5           0     3  ...     13        8570   

      sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
7252           3480      1910          1987    98102  47.6298 -122.323   

      sqft_living15  sqft_lot15  
7252           3940        8800  

[1 rows x 21 columns]

**Resposta** A casa do ID 6762700020 tem o maior valor de venda de 7700000.0 

###  5 - Qual a casa com o maior número de quartos?

In [15]:
df.sort_values(by='bedrooms',ascending=False).head(1)

id             date     price  bedrooms  bathrooms  \
15870  2402100895  20140625T000000  640000.0        33       1.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
15870         1620      6000     1.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
15870        1040            580      1947             0    98103  47.6878   

          long  sqft_living15  sqft_lot15  
15870 -122.331           1330        4700  

[1 rows x 21 columns]

**Resposta** A maior com numeros de quartos e a casa de ID 2402100895 tem 33 quartos.

### 6 - Qual a soma total de quartos do conjunto de dados? 

In [16]:
df['bedrooms'].sum()

72273

**Resposta** O total de quartos do conjunto de dados e 72273

###  7 - Quantas casas possuem 2 banheiros?

In [64]:
#use o loc pra contar na coluna bathroom quanto sao igual a 2
df.loc[df['bathrooms'] == 2].shape[0]


1913

**Resposta** 1913 casa possuem 2 banheiros.

### 8 -  Qual o preço médio de todas as casas no conjunto de dados?

In [18]:
# use o .mean() para calcular o preco medio
df['price'].mean()

541649.962726255

**Resposta** Opreco medio das casa e de 541649.96

### 9 - Qual o preço médio de casas com 2 banheiros? 

In [19]:
# use o .mean() com o .loc para calcular o preco medio de casas com 2 banheiros
df[['price','bathrooms']].loc[df['bathrooms'] == 2].mean()

price        459307.013591
bathrooms         2.000000
dtype: float64

**Resposta** O preco medio de casas com 2 banheiros e de 459307.01

###  10 - Qual o preço mínimo entre as casas com 3 quartos?

In [79]:
df[['price','bedrooms']].loc[df['bedrooms'] == 3].min()

price       89000.0
bedrooms        3.0
dtype: float64

**Resposta** Opreco medio de casas com 3 quartos e de 89000.0

### 11 -  Quantas casas possuem mais de 300 metros quadrados de sala de estar?

Para responder essa questão, será necessário fazer a conversão de pés para metros quadrados, bastando, para tanto, realizar a multiplicação por 0.0929.

In [21]:
# criando uma nova coluna para receber a conversão
df['m2_living'] = df['sqft_living'] * 0.092

# verificando o resultado
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...        1180              0   
1      7242     2.0           0     0  ...        2170            400   
2     10000     1.0           0     0  ...         770              0   
3      5000     1.0           0     0  ...        1050            910   
4      8080     1.0           0     0  ...        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  m2_living  
0        5650     108.56  
1        7639     236.44  
2        8062      70.84  
3        5000     180.32  
4        7503     154.56  

[5 rows x 22 columns]

In [22]:
# use o .shape[] para contar as linhas na coluna m2_living com valores maior que 300
df [df ['m2_living'] > 300] .shape[0]

2135

**Resposta** 2251 casas possuem sala de estar com mais de 300 m2

### 12 -  Quantas casas tem mais de 2 andares?

In [23]:
#use o .shape para contar na coluna Floor
df[df['floors'] > 2] .shape[0]

780

**Resposta** 780 casas possuem masi de 2 andares.

###  13 - Quantas casas tem vista para o mar?

In [24]:
# use o != 'diferente' e o .shape para contar os valores
df[df['waterfront'] != 0] .shape[0]

163

**Resposta** 163 Casas tem vista para o mar.

### 14 -  Das casas com vista para o mar, quantas tem 3 quartos?

In [25]:
# use o & para fazer o and e juntar os dicionario entre parentese "()"
df[(df['waterfront'] != 0) & (df['bedrooms'] == 3)] .shape[0]

64

**Resposta** 64 Casas tem vista para o mar e possuem 3 quartos.

### 15 -  Das casas com mais de 300 metros quadrados de sala de estar, quantas tem mais de 2 banheiros?

In [26]:
# use o & para fazer o and e juntar os dicionario entre parentese "()"
df[(df['m2_living'] > 300) & (df['bathrooms'] > 2)] . shape[0]

2082

**Resposta** 2082 casas tem mais de 300 m2 de sala e masi de  2 banheiros

In [27]:
df[(df['price'] > 3000000) & (df['bathrooms'] > 2)] . shape[0]

44

# Check point to Export Dataset 1.0

**Criando copia do dataset modificado para visualizacao no POWER BI**

In [37]:
dfv = df
dfv.to_csv('data/kc_house_data_viz.csv', index=False)

In [38]:
dfv.head()

id       date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520 2014-10-13  221900.0         3       1.00         1180   
1  6414100192 2014-12-09  538000.0         3       2.25         2570   
2  5631500400 2015-02-25  180000.0         2       1.00          770   
3  2487200875 2014-12-09  604000.0         4       3.00         1960   
4  1954400510 2015-02-18  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...        1180              0   
1      7242     2.0           0     0  ...        2170            400   
2     10000     1.0           0     0  ...         770              0   
3      5000     1.0           0     0  ...        1050            910   
4      8080     1.0           0     0  ...        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  m2_living  
0        5650     108.56  
1        7639     236.44  
2        8062      70.84  
3        5000     180.32  
4        7503     154.56  

[5 rows x 22 columns]